### ISSUE TITLE
ASN hierarchies and the pain of removing "local-as" from netixlan

### ISSUE THREAD

**Issue by eb0815** [peeringdb/peeringdb#635]
HI guys,

Arnold encouraged me to at least come forward with this, though I still believe we're a special case.

We are present at dozens of IXPs world-wide, and while we originate from the same ASN everywhere, we also use a different local ASN for peering locally at each IXP.

Up until two weeks ago, the pdb web page for the master ASN "net" object would give you this, PLUS all the IXPs where we peer through local ASNs. As of now, this is gone completely (I can only see the ASNs from the ORG record, but without any details.

What I'm trying to achieve:

  - one master ASN with local sub-ASNs, so I can direct potential peering partners and interested parties to the one NET object (e.g, a peeringdb link)
  - get "peerfinder.py" working again, since now it shows basically nothing anymore

I'm not sure these concerns were addressed when deciding to remove the local ASN bit from netixlan, and I understand why you wanted to get rid of it.

So - is there a way to achieve the above (this is basically a hierarchical setup)?

Cheers,
Elmar.

**Comment by grizz** [comment]
Absolutely thanks for coming forward! We reached out to quite a few people and hadn't heard of anyone using this use case.

It is currently be supported by having each ASN as a separate network under your parent organization, and probably linking to the master ASN in the notes section. Without having seen `peerfinder.py`, I would imagine that could be fixed to use the ASN you use at each IXP fairly easily.

Would that work for you? Otherwise, if you want to give me your ASNs, that might help, feel free to email me privately and we can try to figure out a good solution.

**Comment by arnoldnipper** [comment]
> It is currently be supported by having each ASN as a separate network under your parent organization, and probably linking to the master ASN in the notes section.

The `notes` field is unstructured text and not very well suited to be eaten by robots. I guess, Elmar is looking for something like a `master_asn` field which points to the ASN which holds all the policies. That way you only would have to maintain one `net` object. I also know that ISC's need might be a corner case, but it's worth to look at it.

We still have a few left-overs from PDB1.0. The concept of primary ASN is one of them. Also having `local_asn` in the `netfac` object instead `asn` is another one. Maybe we should put up some brain for V3 how to handle this efficiently. I.e. basically an organisation would like to show all of its interconnections (`netixlan`) and presences (`netfac`) in **one** place.

**Comment by mcmanuss8** [comment]
I like the idea of showing an organization's IXes and Facilities. Right now /org is a bit barebones, so adding more data there seems like a good thing.

Then, once this is built, if an organization has multiple networks, we also flag that somehow on the /net page s.t. it's more obvious to the user that they might want to look at the the /org page to see the bigger picture. We suffer from this at https://www.peeringdb.com/org/9660 so I would definitely like to see this done. 

So, proposal:

1) Add "Public Peering Exchange Points" widget on /org that shows the set of netixlan values for *all* the networks in the org.
2) Ditto for "Private Peering Facilities", below the above.
3) On /net, add a red flag between the heading and filter box if that org has multiple networks with exchanges and/or facilities. Tooltip on the flag should say something like "This network is part of an organization that has multiple ASNs, that may be present in other exchanges. Click here to see all of them" (and similar wording w/ s/exchanges/facilities/g for the facilities box). Make the flag link to /org so you can see 1 and 2, above.

**Comment by arnoldnipper** [comment]
> 1. Add "Public Peering Exchange Points" widget on /org that shows the set of netixlan values for _all_ the networks in the org.

+1 however, /org might become overloaded as it also show the facilities and the ixes an organisation owns. Now it would also have to show all netixlan and netfac, and in case the org runs ixes also the ixfac. Maybe we should show on the /org level only the org information and the admin UI as well as two widgets, one for the records the org holds and the other one for the operational records (netixlan, netfac, ixfac)

**Comment by mcmanuss8** [comment]
> > 1. Add "Public Peering Exchange Points" widget on /org that shows the set of netixlan values for _all_ the networks in the org.
> 
> +1 however, /org might become overloaded as it also show the facilities and the ixes an organisation owns. Now it would also have to show all netixlan and netfac, and in case the org runs ixes also the ixfac. Maybe we should show on the /org level only the org information and the admin UI as well as two widgets, one for the records the org holds and the other one for the operational records (netixlan, netfac, ixfac)

Good point. Maybe we should think of a worst case /org (one w/ multiple nets, facilities, and netixlan, netfac, ixfac) and mock up what it would look like?

**Comment by arnoldnipper** [comment]
> Good point. Maybe we should think of a worst case /org (one w/ multiple nets, facilities, and netixlan, netfac, ixfac) and mock up what it would look like?

https://peeringdb.com/org/2 ... have fun :)

**Comment by arnoldnipper** [comment]
Hello, newbies @Yo-Robinson, @shane-kerr, and @ynbrthr ... any ideas how to implement this?

**Comment by arnoldnipper** [comment]
> Good point. Maybe we should think of a worst case /org (one w/ multiple nets, facilities, and netixlan, netfac, ixfac) and mock up what it would look like?

@mcmanuss8, to give you an idea. [Equinix](https://peeringdb.com/org/2) is linked to 12,000+ objects

**Comment by eb0815** [comment]
notifications@github.com (Arnold Nipper) wrote:

> > Good point. Maybe we should think of a worst case /org (one w/ multiple nets, facilities, and netixlan, netfac, ixfac) and mock up what it would look like?
>
> @mcmanuss8, to give you an idea. [Equinix](https://peeringdb.com/org/2) is linked to 12,000+ objects

I think I'll pitch in, since I was obviously the firestarter here.
https://www.peeringdb.com/net/12712 is our F-Root "source ASN", but it never
peers directly, always through local ASNs, which formerly hung off 3557, so
3557 "was present at IXPs=netixlans/netfac/ixfac" with a different local
"ASN=net". That got removed, so now people have problems finding common peering
points. I am about to beautify and release a version of peerfinder.py that
traverses from the org object down, which helps a bit.

Removing that feature unfortunately broke a couple things for us which are
still not super easy to deal with.

If you have questions or need something tested, ping me!

Cheers,
    Elmar.

**Comment by arnoldnipper** [comment]
> Good point. Maybe we should think of a worst case /org (one w/ multiple nets, facilities, and netixlan, netfac, ixfac) and mock up what it would look like?

I guess, what the users are interested in is to see all of the `netixlan` and `netfac` on one page as this answers the question "Where am I able to meet you?". 

That then boils down to 

```tsv
1330    58      84      Internet Systems Consortium, Inc. (ISC)
396     1       492     Hurricane Electric
393     2       405     Packet Clearing House
9660    2       403     Facebook
8896    3       400     Microsoft Corporation
574     8       392     Google LLC
4715    1       383     Cloudflare
14      3       374     Akamai Technologies
1520    2       316     Amazon.com
365     1       305     Cogent Communications, Inc.
2       36      283     Equinix
```

where the 1st column is the `org_id`, the 2nd the number of networks the `org` owns, the 3rd the combined number of `netfac`and `netixlan` and the 4th column the name of the organisation. That's doable on one page (we already do this for HE, Cloudflare and Cogent)

We could then e.g. have on https://peeringdb.com/org/xxxx always a first entry under "Networks" which reads "Your combined networks" and links to a new page with all the networks on the left-hand side and all the "Public Peering Exchange Points" and "Private Peering Facilities" on the right-hand side.  From each /net you can also easily link to this page as you have `org_id` at hand anyway.

@pc and @eb0815 what do you think about it?

**Comment by leovegoda** [comment]
Is the concern here that very long lists will be unworkable for some users? If that's the case, couldn't we offer users an option to paginate lists, with 10, 25, 50 or whatever the appropriate number of entries they want is per page?

**Comment by arnoldnipper** [comment]
> Is the concern here that very long lists will be unworkable for some users? If that's the case, couldn't we offer users an option to paginate lists, with 10, 25, 50 or whatever the appropriate number of entries they want is per page?

That was a concern. However, real figures show that we will stay workable all the time.  And for displaying we are using a fixed window size and a scroll bar. IMO the question now only is, where to implement the aggregated view

**Comment by eb0815** [comment]
Re Arnold, Leo,

notifications@github.com (Arnold Nipper) wrote:

> I guess, what the users are interested in is to see all of the netixlan and
> netfac on one page as this answers the question "Where am I able to meet you?".
...

> @pc and @eb0815 what do you think about it?

We're not interested in what it looks like on the web page, nobody uses that to
find common locations and fabrics. The problem that popped up was that the tools
that people use - peerfinder.py - stopped working for anything that involved ASNs
that don't peer directly.

I understand that you want(ed) to move the connection from the ASN to the org
database object; understandable from a DB maintainer pov, yet, now the sanest
thing to do would be to add org objects to keep the ASNs sorted. For us, this
is only two, one for ISC proper (aka AS1280), one for F-Root (aka AS3557 and
a heap of peering ASNs). For others it might be a whole lot more, if they want
to keep things organised.


Yours,
	Elmi.

-- 

"Machen Sie sich erst einmal unbeliebt. Dann werden Sie auch ernstgenommen."
							     (Konrad Adenauer)

--------------------------------------------------------------[ ELMI-RIPE ]---

**Comment by arnoldnipper** [comment]
@eb0815, if you are not longer interested in an implementation, please let us know.

**Comment by eb0815** [comment]
Hey Arnold, folks,

if it is easy and no-hassle for you to implement this, it would make us very happy. If not, we can live with it, as we did for the past year. 

Cheery-up,
Elmar
